### PR TITLE
[LICENS-45]: Token auth

### PR DESCRIPTION
**Pull request by defunctl**
## Primary Changes
- Adds token authorization flow, see the Updated [README.md](https://github.com/stellarwp/uplink/blob/29306946e41fcc21d3c8a44801e62789acc55705/README.md#license-authorization) for details.
- Adds a V3 API client to talk to StellarWP Licensing API V3.
- Tests now run a set using multisite and singlesite, e.g. `wpunit --env singlesite` and `wpunit --env multisite`

## Additional Changes
- Removes `composer.lock` because this is a library and loosens restrictions of dependencies. Adds `phpspec/prophecy` for mocking as it's no longer automatically included in later versions of PHPUnit, and existing tests relied on this.
- Removes auto generated Codeception files that were committed. 

## Demo
- https://d.pr/v/huH5E7 (same as the other PR's).

## Companion PR's
- https://github.com/stellarwp/licensing/pull/51
- https://github.com/stellarwp/uplink-origin/pull/46

### CODE REVIEW

File: src/Uplink/Admin/Notice.php [hunk 1]
```diff
@@ -16,12 +16,12 @@ class Notice {
 	/**
 	 * @var array<mixed>
 	 */
-	protected array $saved_notices = [];
```
Comment (defunctl) [inline src/Uplink/Admin/Notice.php:19]: These were added previously by another developer, technically this project does not support the minimum of [PHP7.4](https://github.com/stellarwp/uplink/blob/29306946e41fcc21d3c8a44801e62789acc55705/composer.json#L31)

File: src/Uplink/Register.php [hunk 2]
```diff
@@ -11,34 +11,34 @@ class Register {
...
 	 * @param string $version Resource version.
 	 * @param string $path Resource path to bootstrap file.
 	 * @param string $class Resource class.
 	 * @param string $license_class Resource license class
 	 *
 	 * @return Resources\Resource
 	 */
-	public static function plugin( $name, $slug, $version, $path, $class, $license_class = null ) {
-		return Resources\Plugin::register( $name, $slug, $version, $path, $class, $license_class );
+	public static function plugin( $slug, $name, $version, $path, $class, $license_class = null ) {
```
Comment (defunctl) [inline src/Uplink/Register.php:23]: These variables were in the correct order, but they were named for the wrong things, so I swapped them.

If you follow the chain this creates, all the `register()` methods take a slug first and this matches the current documentation in the README.

e.g.

```php
Register::plugin(
	$plugin_slug,
	$plugin_name,
	$plugin_version,
	$plugin_path,
	$plugin_class,
	$license_class // This is optional.
);

Register::service(
	$service_slug,
	$service_name,
	$service_version,
	$plugin_path,
	$plugin_class
);
```

File: src/Uplink/Auth/Authorizer.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * Determines if the current site will allow the user to use the authorize button.
+ */
+final class Authorizer {
+
+	/**
+	 * @var Pipeline
+	 */
+	private $pipeline;
+
+	public function __construct( Pipeline $pipeline ) {
```
Comment (borkweb) [inline src/Uplink/Auth/Authorizer.php:17]: Docblock

File: src/Uplink/Auth/Nonce.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
+<?php declare( strict_types=1 );
+
+namespace StellarWP\Uplink\Auth;
+
+use StellarWP\Uplink\Config;
+
+final class Nonce {
+
+	public const NONCE_SUFFIX = '_uplink_nonce';
```
Comment (borkweb) [inline src/Uplink/Auth/Nonce.php:9]: Docblock

File: src/Uplink/Auth/Nonce.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		return $nonce;
+	}
+
+	public function create_url( string $url ): string {
+		$url = str_replace( '&amp;', '&', $url );
+
+		return esc_html( add_query_arg( '_uplink_nonce', $this->create(), $url ) );
+	}
+
+	private function key(): string {
```
Comment (borkweb) [inline src/Uplink/Auth/Nonce.php:65]: Docblocks

File: src/Uplink/Auth/Provider.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+			}
+		);
+
+		$this->register_nonce();
+		$this->register_authorizer();
+		$this->register_auth_disconnect();
+		$this->register_auth_connect();
+	}
+
+	private function register_nonce(): void {
```
Comment (borkweb) [inline src/Uplink/Auth/Provider.php:38]: Docblock

File: src/Uplink/Auth/Provider.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		add_action( 'admin_init', function(): void {
+			$this->container->get( Disconnect_Controller::class )->maybe_disconnect();
+		}, 9, 0 );
+	}
+
+	private function register_auth_connect(): void {
+		add_action( 'admin_init', function(): void {
+			$this->container->get( Connect_Controller::class )->maybe_store_token_data();
+		}, 9, 0 );
+	}
```
Comment (borkweb) [inline src/Uplink/Auth/Provider.php:90]: Two things here.

1. Docblocks
2. To be a good WordPress citizen, these actions need to be unhookable, so we cannot use anonymous functions here. While we feel that this logic belongs in `admin_init`, an end user of our plugins might have a weird situation where that isn't viable, so providing a situation where they can unhook and re-hook to the action of their choosing is important.

File: src/Uplink/Auth/Token/Connector.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	 * @var Authorizer
+	 */
+	private $authorizer;
+
+	/**
+	 * @var Token_Manager
+	 */
+	private $token_manager;
+
+	public function __construct(
```
Comment (borkweb) [inline src/Uplink/Auth/Token/Connector.php:21]: Docblock

File: src/Uplink/Auth/Token/Contracts/Token_Manager.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	public const TOKEN_SUFFIX = 'uplink_auth_token';
+
+	/**
+	 * Returns the option_name/network_option_name that is used to store tokens.
+	 *
+	 * @return string
+	 */
+	public function option_name(): string;
+
+
```
Comment (borkweb) [inline src/Uplink/Auth/Token/Contracts/Token_Manager.php:22]:
```suggestion
```

File: src/Uplink/Auth/Token/Disconnector.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	 * @var Authorizer
+	 */
+	private $authorizer;
+
+	/**
+	 * @var Token_Manager
+	 */
+	private $token_manager;
+
+	public function __construct(
```
Comment (borkweb) [inline src/Uplink/Auth/Token/Disconnector.php:20]: Docblock

File: src/Uplink/Auth/Token/Token_Manager.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	/**
+	 * The option name to store the token in wp_options table.
+	 *
+	 * @see Config::set_token_auth_prefix()
+	 *
+	 * @var string
+	 */
+	protected $option_name;
+
+	public function __construct( string $option_name ) {
```
Comment (borkweb) [inline src/Uplink/Auth/Token/Token_Manager.php:24]: Docblock

File: src/Uplink/Auth/Token/Token_Manager.php [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		if ( ! $option_name ) {
+			throw new InvalidArgumentException( 'You must set a token prefix with StellarWP\Uplink\Config::set_token_auth_prefix() before using the token manager.' );
```
Comment (borkweb) [inline src/Uplink/Auth/Token/Token_Manager.php:26]:
```suggestion
			throw new InvalidArgumentException( __( 'You must set a token prefix with StellarWP\Uplink\Config::set_token_auth_prefix() before using the token manager.', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Components/Admin/Authorize_Button_Controller.php [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+	/**
+	 * The auth URL for the origin as fetched remotely from the
+	 * licensing server when we want to render the button.
+	 *
+	 * @var string
+	 */
+	private $auth_url = '';
+
+	public function __construct(
```
Comment (borkweb) [inline src/Uplink/Components/Admin/Authorize_Button_Controller.php:47]: Docblock

File: src/Uplink/Components/Admin/Authorize_Button_Controller.php [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+	 *
+	 * @throws InvalidArgumentException
+	 */
+	public function render( array $args = [] ): void {
+		global $pagenow;
+
+		$slug = $args['slug'] ?? '';
+
+		if ( empty ( $slug ) ) {
+			throw new InvalidArgumentException( 'The Product slug cannot be empty' );
```
Comment (borkweb) [inline src/Uplink/Components/Admin/Authorize_Button_Controller.php:77]:
```suggestion
			throw new InvalidArgumentException( __( 'The Product slug cannot be empty', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Config.php [hunk 15]
```diff
@@ -27,15 +34,15 @@ class Config {
...
 	 * @since 1.0.0
 	 *
+	 * @throws RuntimeException
+	 *
+	 * @return ContainerInterface
 	 */
 	public static function get_container() {
 		if ( self::$container === null ) {
-			throw new \RuntimeException( 'You must provide a container via StellarWP\Uplink\Config::set_container() before attempting to fetch it.' );
+			throw new RuntimeException( 'You must provide a container via StellarWP\Uplink\Config::set_container() before attempting to fetch it.' );
```
Comment (borkweb) [inline src/Uplink/Config.php:45]: Oops. Not having exceptions in this class file translated before was an oversight.

```suggestion
			throw new RuntimeException( __( 'You must provide a container via StellarWP\Uplink\Config::set_container() before attempting to fetch it.', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Config.php [hunk 16]
```diff
@@ -48,9 +55,9 @@ public static function get_container() {
 	 *
 	 * @return string
 	 */
-	public static function get_hook_prefix() {
+	public static function get_hook_prefix(): string {
 		if ( self::$hook_prefix === null ) {
-			throw new \RuntimeException( 'You must provide a hook prefix via StellarWP\Uplink\Config::set_hook_prefix() before attempting to fetch it.' );
+			throw new RuntimeException( 'You must provide a hook prefix via StellarWP\Uplink\Config::set_hook_prefix() before attempting to fetch it.' );
```
Comment (borkweb) [inline src/Uplink/Config.php:60]:
```suggestion
			throw new RuntimeException( __( 'You must provide a hook prefix via StellarWP\Uplink\Config::set_hook_prefix() before attempting to fetch it.', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Config.php [hunk 17]
```diff
@@ -63,9 +70,9 @@ public static function get_hook_prefix() {
 	 *
 	 * @return string
 	 */
-	public static function get_hook_prefix_underscored() {
+	public static function get_hook_prefix_underscored(): string {
 		if ( self::$hook_prefix === null ) {
-			throw new \RuntimeException( 'You must provide a hook prefix via StellarWP\Uplink\Config::set_hook_prefix() before attempting to fetch it.' );
+			throw new RuntimeException( 'You must provide a hook prefix via StellarWP\Uplink\Config::set_hook_prefix() before attempting to fetch it.' );
```
Comment (borkweb) [inline src/Uplink/Config.php:75]:
```suggestion
			throw new RuntimeException( __( 'You must provide a hook prefix via StellarWP\Uplink\Config::set_hook_prefix() before attempting to fetch it.', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Config.php [hunk 18]
```diff
@@ -115,7 +126,44 @@ public static function set_container( ContainerInterface $container ) {
...
+	 *
+	 * @param  string  $prefix
+	 *
+	 * @throws RuntimeException|InvalidArgumentException
+	 *
+	 * @return void
+	 */
+	public static function set_token_auth_prefix( string $prefix ): void {
+		if ( ! self::has_container() ) {
+			throw new RuntimeException( 'You must set a container with StellarWP\Uplink\Config::set_container() before setting a token auth prefix.' );
```
Comment (borkweb) [inline src/Uplink/Config.php:148]:
```suggestion
			throw new RuntimeException( __( 'You must set a container with StellarWP\Uplink\Config::set_container() before setting a token auth prefix.', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Config.php [hunk 19]
```diff
@@ -115,7 +126,44 @@ public static function set_container( ContainerInterface $container ) {
...
+		$prefix = Sanitize::sanitize_title_with_hyphens( rtrim( $prefix, '_' ) );
+		$key    = sprintf( '%s_%s', $prefix, Token_Manager::TOKEN_SUFFIX );
+
+		// The option_name column in wp_options is a varchar(191)
+		$max_length = 191;
+
+		if ( strlen( $key ) > $max_length ) {
+			throw new InvalidArgumentException(
+				sprintf(
+					'The token auth prefix must be at most %d characters, including a trailing hyphen.',
```
Comment (borkweb) [inline src/Uplink/Config.php:160]:
```suggestion
					__( 'The token auth prefix must be at most %d characters, including a trailing hyphen.', '%TEXTDOMAIN%' ),
```

File: src/Uplink/Notice/Notice.php [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+	private $alt;
+
+	/**
+	 * Whether this should be a large notice.
+	 *
+	 * @var bool
+	 */
+	private $large;
+
+	public function __construct(
```
Comment (borkweb) [inline src/Uplink/Notice/Notice.php:61]: Docblock

File: src/Uplink/Notice/Notice.php [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+		string $type,
+		string $message,
+		bool $dismissible = false,
+		bool $alt = false,
+		bool $large = false
+	) {
+		if ( ! in_array( $type, self::ALLOWED_TYPES, true ) ) {
+			throw new InvalidArgumentException( sprintf(
+					'Notice $type must be one of: %s',
```
Comment (borkweb) [inline src/Uplink/Notice/Notice.php:70]:
```suggestion
					__( 'Notice $type must be one of: %s', '%TEXTDOMAIN%' ),
```

File: src/Uplink/Notice/Notice.php [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+					implode( ', ', self::ALLOWED_TYPES ) )
+			);
+		}
+
+		if ( empty( $message ) ) {
+			throw new InvalidArgumentException( 'The $message cannot be empty' );
```
Comment (borkweb) [inline src/Uplink/Notice/Notice.php:76]:
```suggestion
			throw new InvalidArgumentException( __( 'The $message cannot be empty', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/Notice/Provider.php [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+	 */
+	public function register(): void {
+		$this->container->bind( Notice_Controller::class, Notice_Controller::class );
+		$this->container->bind( Notice_Handler::class, static function ( $c ): Notice_Handler {
+			return new Notice_Handler( $c->get( Notice_Controller::class ) );
+		} );
+
+		add_action( 'admin_notices', function (): void {
+			$this->container->get( Notice_Handler::class )->display();
+		}, 12, 0 );
```
Comment (borkweb) [inline src/Uplink/Notice/Provider.php:20]: This cannot be an anonymous function. It needs to be easily unhookable.

File: src/Uplink/Traits/With_Debugging.php [hunk 24]
```diff
@@ -0,0 +1,11 @@
+<?php declare( strict_types=1 );
+
+namespace StellarWP\Uplink\Traits;
+
+trait With_Debugging {
+
+	protected function is_wp_debug(): bool {
```
Comment (borkweb) [inline src/Uplink/Traits/With_Debugging.php:7]: Docblock

File: src/Uplink/Uplink.php [hunk 25]
```diff
@@ -13,19 +16,34 @@ class Uplink {
 	 */
 	public static function init() {
 		if ( ! Config::has_container() ) {
-			throw new \RuntimeException( 'You must call StellarWP\Uplink\Config::set_container() before calling StellarWP\Telemetry::init().' );
+			throw new RuntimeException( 'You must call StellarWP\Uplink\Config::set_container() before calling StellarWP\Telemetry::init().' );
```
Comment (borkweb) [inline src/Uplink/Uplink.php:19]:
```suggestion
			throw new RuntimeException( __( 'You must call StellarWP\Uplink\Config::set_container() before calling StellarWP\Telemetry::init().', '%TEXTDOMAIN%' ) );
```

File: src/Uplink/View/Provider.php [hunk 26]
```diff
@@ -0,0 +1,21 @@
+<?php declare( strict_types=1 );
+
+namespace StellarWP\Uplink\View;
+
+use League\Plates\Engine;
```
Comment (borkweb) [inline src/Uplink/View/Provider.php:5]: I think it is too heavy handed to be introducing a templating engine as a dependency into this uplink library.

File: src/views/admin/authorize-button.php [hunk 27]
```diff
@@ -0,0 +1,27 @@
+<?php declare( strict_types=1 );
+/**
+ * The authorize button view, allowing the user to authorize their install with
+ * the license server via the origin site.
+ *
+ * @see \StellarWP\Uplink\Components\Admin\Authorize_Button_Controller
+ *
+ * @var \League\Plates\Template\Template $this
```
Comment (borkweb) [inline src/views/admin/authorize-button.php:8]: We'll need to remove this.

File: src/views/admin/notice.php [hunk 28]
```diff
@@ -0,0 +1,16 @@
+<?php declare( strict_types=1 );
+/**
+ * Render a WordPress dashboard notice.
+ *
+ * @see \StellarWP\Uplink\Notice\Notice_Controller
+ *
+ * @var \League\Plates\Template\Template $this
```
Comment (borkweb) [inline src/views/admin/notice.php:7]: Remove

File: src/Uplink/Components/Admin/Authorize_Button_Controller.php [hunk 29]
```diff
@@ -0,0 +1,210 @@
+<?php declare( strict_types=1 );
+
+namespace StellarWP\Uplink\Components\Admin;
+
+use InvalidArgumentException;
+use League\Plates\Engine;
```
Comment (borkweb) [inline src/Uplink/Components/Admin/Authorize_Button_Controller.php:6]: We'll need to remove this.

File: src/Uplink/Components/Controller.php [hunk 30]
```diff
@@ -0,0 +1,47 @@
+<?php declare( strict_types=1 );
+
+namespace StellarWP\Uplink\Components;
+
+use League\Plates\Engine;
```
Comment (borkweb) [inline src/Uplink/Components/Controller.php:5]: We'll need to remove this

**Review by borkweb — CHANGES_REQUESTED** [review]
There's a bunch of great stuff in this code! Nice changes. A few notes:

## Anonymous functions for actions

As a whole, we need to avoid using anonymous functions for actions because everything needs to be unhookable / movable. Our default flow that we believe is fine, may not actually be viable on some sites and we need to give those sites the ability to change or manipulate the actions.

## League Plates

We cannot introduce that library into the uplink library. That's too opinionated of a tool for the job. If this was an internal library for our own sites, it'd be one thing. Because it will be used on multiple brand products and then those will be distributed out in the wild, adopting something like this needs a broader discussion.

## Docblocks on constructors

We standardize on having them.

## Translations for exceptions

Historically, WordPress, WooCommerce, TEC, etc did not run exception strings through translate functions. However, the newer code of WP and WC do. I think we should follow suit and support non-english developers by translating Exception strings. There were a few exceptions that were altered (but pre-existing) that I went ahead and added a code suggestion for that adds translation functions as a wrapper.

File: src/Uplink/API/Client.php [hunk 31]
```diff
@@ -25,7 +25,7 @@ class Client {
 	 *
 	 * @var string
 	 */
-	public static $api_root = '/api/plugins/v2/';
+	protected $api_root = '/api/plugins/v2/';
```
[the remark this answers is not in the document]
Comment (defunctl) [reply src/Uplink/API/Client.php:28]: Removed the translations.

[the remark this answers is not in the document]
Comment (defunctl) [reply src/Uplink/API/Client.php:28]: Completely agree and I did the same check :grin:

File: src/Uplink/Auth/Admin/Disconnect_Controller.php [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+	public function maybe_disconnect(): void {
+		if ( empty( $_GET[ self::ARG ] ) || empty( $_GET['_wpnonce'] ) ) {
+			return;
+		}
+
+		if ( ! is_admin() || wp_doing_ajax() ) {
+			return;
+		}
+
+		if ( wp_verify_nonce( $_GET['_wpnonce'], self::ARG ) ) {
```
[the remark this answers is not in the document]
Comment (defunctl) [reply src/Uplink/Auth/Admin/Disconnect_Controller.php:44]: TIL

**Review by borkweb — APPROVED** [review]

**Review by borkweb — APPROVED** [review]

**Pull request merged**